src/common/src/hash/key_v2.rs

```diff
@@ -0,0 +1,389 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+use std::hash::{Hash, Hasher};
+use std::marker::PhantomData;
+
+use bytes::BufMut;
+use educe::Educe;
+use either::{for_both, Either};
+use itertools::Itertools;
+use tinyvec::ArrayVec;
+
+use super::{HeapNullBitmap, NullBitmap, XxHash64HashCode};
+use crate::array::{Array, ArrayBuilder, ArrayBuilderImpl, ArrayImpl, ArrayResult, DataChunk};
+use crate::estimate_size::EstimateSize;
+use crate::for_all_type_pairs;
+use crate::hash::{HashKeyDe, HashKeySer};
+use crate::row::OwnedRow;
+use crate::types::{DataType, Datum, ScalarImpl};
+use crate::util::hash_util::XxHash64Builder;
+use crate::util::iter_util::ZipEqFast;
+
+/// The storage where the hash key resides in memory.
+pub trait KeyStorage: 'static {
+    /// The key type that is used to store the hash key.
+    type Key: AsRef<[u8]> + EstimateSize + Clone + Send + Sync + 'static;
+
+    /// The buffer type that is used to build the hash key.
+    type Buffer: Buffer<Sealed = Self::Key>;
+}
+
+/// Associated type for [`KeyStorage`] used to build the hash key.
+pub trait Buffer: BufMut + 'static {
+    /// The sealed key type.
+    type Sealed;
+
+    /// Returns whether this buffer allocates on the heap.
+    fn alloc() -> bool;
+
+    /// Creates a new buffer with the given capacity.
+    fn with_capacity(cap: usize) -> Self;
+
+    /// Seals the buffer and returns the sealed key.
+    fn seal(self) -> Self::Sealed;
+}
+
+/// Key storage that uses a on-stack buffer and key, backed by a byte array with length `N`.
+///
+/// Used when the encoded length of the hash key is known to be less than or equal to `N`.
+pub struct StackStorage<const N: usize>;
+
+impl<const N: usize> KeyStorage for StackStorage<N> {
+    type Buffer = StackBuffer<N>;
+    type Key = [u8; N];
+}
+
+/// The buffer for building a hash key on a fixed-size byte array on the stack.
+pub struct StackBuffer<const N: usize>(ArrayVec<[u8; N]>);
```

Comment:
Why not directly use array

Reply:
Yep. An `ArrayVec` is exactly an array with a "current length" on which some methods are encapsulated. We use `ArrayVec` simply for clearer implementation of `BufMut`. 🤣